src/Vacuum/Vacuum_math.jl

```diff
@@ -0,0 +1,728 @@
+
+    # n = 0
+    P[1] = P0_minus_half(s)
+    if n == 0
+        return P
+    end
+
+    # n = 1
+    P[2] = P1_minus_half_wrong(s)
+    if n == 1
+        return P
+    end
+
+    # n > 1
+    for i in 1:n-1
+        # Chance eq.(47)
+        P[i+2] = -2 * i * s / sqrt(s^2 - 1) * P[i+1] - (i - 0.5)^2 * P[i]
+    end
+
+    return P
+end
+
+#############################################################
+# Green function eq.(36)~(42). replacing green
+#############################################################
+
+"""
+    green(xs, zs, xt, zt, n)
+
+입력:
+- xs, zs: observation points
+- xt, zt: source points
+- xtp, ztp : ∂X'/∂θ, ∂Z'/∂θ
+- n: mode number
+
+반환:
+- G :   2π𝒢ⁿ(θ,θ')
+- aval :    𝒥 ∇'𝒢ⁿ∇'ℒ
+- aval0:    𝒥 ∇'𝒢⁰∇'ℒ
+"""
+function green(xs, zs, xt, zt, xtp, ztp, n, usechancebugs=false)
+
+    xs2 = xs^2
+    xt2 = xt^2
+    x_minus2 = (xs - xt)^2
+    x_multiple = xs * xt
+    ζ = (zs - zt)
+    ζ2 = ζ^2
+
+    ρ2 = x_minus2 + ζ2
+
+    # Chance eq.(41) ℛ = R
+    R4 = ρ2 * (ρ2 + 4 * x_multiple)
+    R2 = sqrt(R4) 
+    R = sqrt(R2)
+
+    # Chance eq.(42) 𝘴 = s
+    s = (xs2 + xt2 + ζ2) / R2
+    
+    # Legendre functions for 
+    # P⁰ = p0, P¹ = p1, Pⁿ = pn, Pⁿ⁺¹ = pp 
+    if usechancebugs
+        legendre = Pn_minus_half_wrong(s, n)
+    else
+        legendre = Pn_minus_half(s, n+1)
+    end
+    p0 = legendre[1]
+    p1 = legendre[2]
+    pp = legendre[end]
+    pn = legendre[end-1]
+
+    # Chance eq.(40) 2π⅁ⁿ = G
+    gg = 2 * sqrt(π) * gamma(0.5 - n) / R
+    G = gg * pn
+
+    # Chance eq.(44)
+    # coefficient
+    grad_gg = gg / R4
+
+    if usechancebugs == false
+        # ∂Gⁿ/∂X' = dG_dX
+        aval1 = (n * (xs2 + xt2 + ζ2)*(xs2 - xt2 + ζ2) - xt2*(xt2-xs2+ζ2)) * pn
+        aval2 = (2.0 * xt * xs * (xs2-xt2+ζ2)) * pp 
+        dG_dX = grad_gg * (aval1 + aval2) / xt
+        
+        # ∂Gⁿ/∂Z' = dG_dZ
+        aval3 = (2.0 * n + 1.0) * (xs2 + xt2 + ζ2) * pn
+        aval4 = 4.0 * x_multiple * pp
+        dG_dZ = grad_gg * (aval3 + aval4) * ζ
+        
+        # Chance eq.(51) 
+        # 𝒥 ∇'𝒢ⁿ∇'ℒ = aval
+        # ∂X'/∂θ = xtp, ∂Z'/∂θ = ztp
+        aval = -xt * (ztp * dG_dX - xtp * dG_dZ)
+
+        # bval
+        bval = G
+
+        # for 𝓃⩵0,  aval0 = 𝒥 ∇'𝒢⁰∇'ℒ 
+        dG_dX0 = grad_gg * (2.0 * xt * xs * (xs2-xt2+ζ2)) * p1 - xt2*(xt2-xs2+ζ2) * p0 / xt
+        dG_dZ0 = grad_gg * ((xs2 + xt2 + ζ2) * p0 + 4.0 * x_multiple * p1) * ζ
+        aval0 = -xt * (ztp * dG_dX0 - xtp * dG_dZ0)
+    else
+        bval  = -gg*pn
+        aval1 = ( n*(xs2+xt2+ζ2)*(xt2-xs2-ζ2)+xt2*(xm2+ζ2))*pn
+        aval2 = 2.0*xt*xs*(xs2-xt2-ζ2)*pp
+        aval3 = ztp*(aval1+aval2) / xt
+        aval4 = (2.0*n+1.0)*(xp2+ζ2)*pn+4.0*xt*xs*pp
+        aval5 = xtp*(zt-zs)*aval4
+        aval6 =(aval3-aval5) / ( xt*R4)
+        aval = - xt2*aval6 * gg / (2*π)
+        aval0 = ztp*(two*xs*(zm2-ζ2)*aleg1 - xt*(xm2+ζ2)*p0)
+        aval0 = aval0 + xtp*(zt-zs)*(4.0*xt*xs*p1+(xp2+zm2)*p0)
+        aval0 = -aval0*xt / (R4*rR)
```

Comment:
If performance critical, there are lots of repeated FLOPs here that you can avoid.